src/mongo/client/dbclient.cpp

```diff
@@ -1066,15 +1070,175 @@ namespace mongo {
         return runCommand(db.c_str(), b.done(), *info);
     }
 
-    bool DBClientWithCommands::copyDatabase(const string &fromdb, const string &todb, const string &fromhost, BSONObj *info) {
+    string DBClientWithCommands::_authKeyCopyDB(const string &username,
+                                                const string &password,
+                                                const string &nonce)
+    {
+        md5digest d;
+        string passwordDigest = createPasswordDigest( username, password );
+        {
+            md5_state_t st;
+            md5_init(&st);
+            md5_append(&st, (const md5_byte_t *) nonce.c_str(), nonce.size() );
+            md5_append(&st, (const md5_byte_t *) username.data(), username.length());
+            md5_append(&st, (const md5_byte_t *) passwordDigest.c_str(), passwordDigest.size() );
+            md5_finish(&st, d);
+        }
+        return digestToString( d );
+    }
+
+    bool DBClientWithCommands::copyDatabase(const string &fromdb,
+                                            const string &todb,
+                                            const string &fromhost,
+                                            const string &mechanism,
+                                            const string &username,
+                                            const string &password,
+                                            BSONObj *info) {
         BSONObj o;
         if ( info == 0 ) info = &o;
-        BSONObjBuilder b;
-        b.append("copydb", 1);
-        b.append("fromhost", fromhost);
-        b.append("fromdb", fromdb);
-        b.append("todb", todb);
-        return runCommand("admin", b.done(), *info);
+        BSONObjBuilder copydbCmd;
+        copydbCmd.append("copydb", 1);
+        copydbCmd.append("fromhost", fromhost);
+        copydbCmd.append("fromdb", fromdb);
+        copydbCmd.append("todb", todb);
+
+        // If we don't have a username, or if we're copying locally,
+        // just run the command without authenticating
+        if ( username == "" || fromhost == "" ) {
+            return runCommand("admin", copydbCmd.done(), *info);
+        }
+
+        // Otherwise, take or guess the auth mechanism
+        string authMech;
+        if ( strcmp( mechanism.c_str(), "DEFAULT" ) != 0 ) {
+            if ( strcmp( mechanism.c_str(), "MONGODB-CR" ) != 0 &&
+                 strcmp( mechanism.c_str(), "SCRAM-SHA-1" ) != 0 ) {
+                throw UserException( 13422,
+                                     "auth mechanism must be MONGODB-CR or SCRAM-SHA-1" );
+            }
+            authMech = mechanism;
+        }
+#ifdef MONGO_SSL
+        else if ( (static_cast<DBClientBase*>(this))->getMaxWireVersion() >= 3 ) {
```

Comment:
I think this is probably undefined behavior since DBClientBase is not a parent of DBClientWithCommands. I think we should use dynamic_cast here, but I will defer to @acmorrow